### DISCUSSION TITLE
Coerce values of certain types while allowing values of other types

### DISCUSSION THREAD

**Discussion by allen-liaoo** [colinhacks/zod#3797] [General]
Hi,
I'm new to zod. I want to be able to coerce `string` to `number`, but I want to allow the value that is `null` to remain `null`. I'm not sure how to specifically allow `null` to pass through uncoerced because coerce would raise an error.

Without handling `null` values:
```js
z.string().or(z.number()).pipe(z.coerce.number())
```
Option 1:
```js
z.union([
    z.string(), 
    z.number(), 
    z.null()
]).pipe(z.coerce.number())
```
Option 1 doesn't work because Number constructor will be called with either `null`, which is not what I intended. I want `null` to remain `null`.

Option 2:
```js
z.union([
    z.string().or(z.number()).pipe(z.coerce.number()), 
    z.null()
])
```
Option 2 doesn't work because if `null` is encountered, `z.coerce.number()` raises an error.

Any help would be appreciated!

**Comment by BrendanC23** [comment]
You can write your own `transform` function:

```ts
const schema = z.object({
    field: z
        .string()
        .or(z.number())
        .nullable()
        .transform((data, ctx) => {
            if (data === null || typeof data === "number") return data;

            const asNumber = Number(data);

            if (isNaN(asNumber)) {
                ctx.addIssue({
                    code: z.ZodIssueCode.custom,
                    message: `The value for "data" is not valid.`,
                });

                return z.NEVER;
            }

            return asNumber;
        }),
});

type schemaInput = z.input<typeof schema>; // { field: string | number | null }
type schemaOutput = z.output<typeof schema>; // { field: number | null }

```

**Reply by allen-liaoo** [reply]
I see, thank you!
Do you know what the difference is between a ZodEffect and a ZodSchema? like what functions are defined for each? The docs weren't super clear about this.
